bpf/bpf_host.c

```diff
@@ -534,6 +536,19 @@ handle_ipv4(struct __ctx_buff *ctx, __u32 secctx,
 		if (ep->flags & ENDPOINT_F_HOST)
 			return CTX_ACT_OK;
 
+		/* add L2 header for L2-less interface, such as cilium_wg0 */
+		ret = maybe_add_l2_hdr(ctx, ep->ifindex, &l2_hdr_required);
+		if (ret != 0)
+			return ret;
+		if (l2_hdr_required && ETH_HLEN == 0) {
+			/* l2 header is added */
+			l3_off += __ETH_HLEN;
+			if (!____revalidate_data_pull(ctx, &data, &data_end,
+						      (void **)&ip4, sizeof(*ip4),
+						      false, l3_off))
+				return DROP_INVALID;
+		}
+
```

Comment:
Should we only include this section when `#ifdef ENABLE_HOST_ROUTING`?

I believe we also need the same handling in the IPv6 path.

Reply:
I think currently ENABLE_HOST_ROUTING can't go along with ipv6, can it? 
host routing requires bpf masquerading: https://github.com/cilium/cilium/blob/master/daemon/cmd/kube_proxy_replacement.go#L461
bpf masquerading isn't supported for ipv6: https://github.com/cilium/cilium/blob/master/daemon/cmd/daemon_main.go#L1413

Reply:
Ah of course - thanks! (it's actually in the works at the moment: https://github.com/cilium/cilium/pull/23165)

So either we add the datapath code already today (knowing that it's unused). Or we add a check/TODO at the right location in the agent code, to make sure we don't forget doing so when support for IPv6 BPF-Masq is added.